src/vm/fieldmarshaler.cpp

```diff
@@ -3364,14 +3364,22 @@ VOID FieldMarshaler_FixedStringAnsi::UpdateNativeImpl(OBJECTREF* pCLRValue, LPVO
             nc = m_numchar - 1;
 
         int cbwritten = InternalWideToAnsi(pString->GetBuffer(),
-                                           nc,
-                                            (CHAR*)pNativeValue,
-                                           m_numchar,
-                                           m_BestFitMap,
-                                           m_ThrowOnUnmappableChar);
+            nc,
+            (CHAR*)pNativeValue,
+            m_numchar,
+            m_BestFitMap,
+            m_ThrowOnUnmappableChar);
+
+        // Handle the case where SizeConst == Number of bytes.For single byte chars 
+        // this will never be the case since nc >= m_numchar check will truncate the last 
+        // character, but for multibyte chars nc>= m_numchar check won't truncate since GetStringLength
+        // gives number of characters but not the actual number of bytes. For such cases need to make 
+        // sure that we dont write one past the buffer.
+        if (cbwritten == m_numchar)
+            --cbwritten;
```

Comment:
Either way, this is breaking change for the border case. 
  No , this's not breaking change for single byte characters.
Why not simply pass m_numchar - 1
  That won't work for single byte char , for example SizeConst=4 and i am marshaling "ab" , there's no need to truncate 'b'  . 

I have considered all alternatives and came to the conclusion that the current implementation is the one with minimum compact impact as it's limited to  just multi-byte character case when SizeConst ==  Number of bytes.